### PR TITLE
Improve performance on Arm64

### PR DESCRIPTION
**Pull request by Developer-Ecosystem-Engineering**
This patch will
- Improve HAL primitives
  - reduction (sum, min, max, sad)
  - signmask
  - mul_expand
  - check_any / check_all

Results on a M1 Macbook Pro

 ```
% ./build-run
cc -march=armv8.4-a+dotprod -O3 -o neon-perf-test-O3 neon-perf-test.c
./neon-perf-test-O3 1000000
CONFIG:
  BUFFER SIZE : 16000
  VECTORS     : 1000
  ITERATIONS  : 1000000
  TOTAL SIZE  : 16000000000
                  Warmup : 75.268 ms

reduce_sum uint8x16:
                  OpenCV : 396.690 ms
              vaddlvq_u8 : 375.593 ms 1.056x

reduce_sum int8x16:
                  OpenCV : 395.530 ms
              vaddlvq_s8 : 375.733 ms 1.053x

reduce_sum uint16x8:
                  OpenCV : 321.458 ms
             vaddlvq_u16 : 320.453 ms 1.003x

reduce_sum int16x8:
                  OpenCV : 321.622 ms
             vaddlvq_s16 : 320.413 ms 1.004x

reduce_sum float32x4:
                  OpenCV : 396.202 ms
              vaddvq_f32 : 321.988 ms 1.230x

reduce_sum4 float32x4:
                  OpenCV : 218.655 ms
              vpaddq_f32 : 121.952 ms 1.793x

signmask uint8x16:
                  OpenCV : 550.870 ms
                   vtbl1 : 395.717 ms 1.392x

signmask uint16x8:
                  OpenCV : 507.869 ms
             vaddlvq_u16 : 320.548 ms 1.584x

signmask uint64x2:
                  OpenCV : 592.592 ms
              vaddvq_u64 : 320.520 ms 1.849x

check_all uint8x16:
                  OpenCV : 443.745 ms
               vminvq_u8 : 375.694 ms 1.181x

check_all uint16x8:
                  OpenCV : 443.904 ms
              vminvq_u16 : 375.623 ms 1.182x

check_all uint32x4:
                  OpenCV : 444.041 ms
              vminvq_u32 : 375.776 ms 1.182x

check_any uint8x16:
                  OpenCV : 403.869 ms
               vmaxvq_u8 : 375.780 ms 1.075x

check_any uint16x8:
                  OpenCV : 403.862 ms
              vmaxvq_u16 : 375.621 ms 1.075x

check_any uint32x4:
                  OpenCV : 403.871 ms
              vmaxvq_u32 : 375.621 ms 1.075x

vmul_expand uint8x16:
                  OpenCV : 202.145 ms
 vmull_u8, vmull_high_u8 : 202.269 ms 0.999x  (better codegen on gcc, although run here is done with clang)

vmul_expand int8x16:
                  OpenCV : 202.147 ms
 vmull_s8, vmull_high_s8 : 202.143 ms 1.000x  (better codegen on gcc, although run here is done with clang)

v_dotprod_expand uint8x16:
                  OpenCV : 512.093 ms
               vdotq_u32 : 202.430 ms 2.530x

v_dotprod_expand_fast uint8x16:
                  OpenCV : 278.198 ms

```
### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

```
force_builders=linux,docs,Android armeabi-v7a,ARMv7,ARMv8
```

### CODE REVIEW

File: cmake/platforms/OpenCV-Darwin.cmake [hunk 1]
```diff
@@ -1 +1,3 @@
-# empty
+if(CMAKE_SYSTEM_PROCESSOR MATCHES "arm64.*|ARM64.*")
```
Comment (tomoaki0705) [inline cmake/platforms/OpenCV-Darwin.cmake:1]: I'm not sure if we should directly add "Darwin" for this purpose.
From the perspective of "Apple M1", yes, this is the straightway.
On the other hand, those instructions are available under complete Arm compatibility.
Thus, this is NOT Apple M1 specific.

We already have [CPU Dispatch](https://github.com/opencv/opencv/blob/master/cmake/OpenCVCompilerOptimizations.cmake) mode implemented, and I highly recommend leveraging this mode.

Here's my suggestion.

- [ ] Remove this file, at least for this PR purpose
- [ ] Add corresponding compiler flags in to https://github.com/opencv/opencv/blob/master/cmake/OpenCVCompilerOptimizations.cmake

File: modules/core/include/opencv2/core/hal/intrin_neon.hpp [hunk 2]
```diff
@@ -78,8 +78,15 @@ CV_CPU_OPTIMIZATION_HAL_NAMESPACE_BEGIN
 #define CV_NEON_AARCH64 0
 #endif
 
-// TODO
+// Check if dot product intrinsics are supported as indicated by the
+// Arm C Language Extensions (ACLE) [1].
+//
+// [1] https://developer.arm.com/documentation/101028/0012/13--Advanced-SIMD--Neon--intrinsics
+#if __ARM_FEATURE_DOTPROD
```
Comment (tomoaki0705) [inline modules/core/include/opencv2/core/hal/intrin_neon.hpp:85]: `#ifdef __ARM_FEATURE_DOTPROD`

File: modules/core/include/opencv2/core/hal/intrin_neon.hpp [hunk 3]
```diff
@@ -537,36 +544,61 @@ OPENCV_HAL_IMPL_NEON_MUL_SAT(v_uint16x8, v_uint32x4)
 inline void v_mul_expand(const v_int8x16& a, const v_int8x16& b,
                          v_int16x8& c, v_int16x8& d)
 {
+#if CV_NEON_AARCH64
+    c.val = vmull_s8(vget_low_s8(a.val), vget_low_s8(b.val));
```
Comment (tomoaki0705) [inline modules/core/include/opencv2/core/hal/intrin_neon.hpp:548]: Please, leverage the existing code.
This line starting with `c.val` is completely identical.
Let's make the scope of `#ifdef` small as possible.

Same suggestion happens to many places.

File: modules/core/include/opencv2/core/hal/intrin_neon.hpp [hunk 4]
```diff
@@ -78,8 +78,15 @@ CV_CPU_OPTIMIZATION_HAL_NAMESPACE_BEGIN
 #define CV_NEON_AARCH64 0
 #endif
 
-// TODO
+// Check if dot product intrinsics are supported as indicated by the
+// Arm C Language Extensions (ACLE) [1].
+//
+// [1] https://developer.arm.com/documentation/101028/0012/13--Advanced-SIMD--Neon--intrinsics
+#if __ARM_FEATURE_DOTPROD
+#define CV_NEON_DOT 1
```
Comment (tomoaki0705) [inline modules/core/include/opencv2/core/hal/intrin_neon.hpp:86]: This line, defining macro `CV_NEON_DOT` is the ONLY thing that is done for M1.
Other things are NOT specific for M1, just not well optimized Armv8 NEON intrinsics.

So honestly, I prefer to separate this PR in 2 different patches.
Enabling M1 specific, following my previous comment, and the other huge modification for general Armv8.

**Review by tomoaki0705 — COMMENTED** [review]
Great contribution.

Still, if this PR is supposed to be M1 specific modification, I think it should be separated into 2 different PR (please see my comment)
It's doing more than it says.

Lastly, not only for this PR, but for anyone who is intended to write SIMD optimization, please leverage using existing performance test code, rather than using your own well-written-comparison-test-code like `neon-perf-test.c`
There is already a test code that covers the performance measurement, and please STOP using your own code.
I really appreciate it if you could consider leveraging existing test code since that will make it way more easier for other contributors to test locally.

**Comment by Developer-Ecosystem-Engineering** [comment]
Thank you for the feedback we will review and come back

File: modules/core/include/opencv2/core/hal/intrin_neon.hpp (hunk 4 above)
Comment (Developer-Ecosystem-Engineering) [reply modules/core/include/opencv2/core/hal/intrin_neon.hpp]: We've provided an updated PR with these changes. We will make a new PR based on this changeset that will make the CV_NEON_DOT changes for M1.

File: modules/core/src/system.cpp [hunk 5]
```diff
@@ -584,6 +585,9 @@ struct HWFeatures
         have[CV_CPU_FP16] = true;
     #endif
     #endif
+    #if CV_DOTPROD
+        have[CV_CPU_DOTPROD] = true;
```
Comment (tomoaki0705) [inline modules/core/src/system.cpp:589]: Unfortunately, this flag has to be checked more carefully.
`CV_DOTPROD` only means that the compiler is capable of dot product instruction and the user wishes to enable this instruction.
What this flag is supposed to express is whether the CPU supports dot product instruction.

Another unfortunate thing is that till today, OpenCV only enabled NEON and FP16 instructions, which are NOT extended instructions in Aarch64, so this could be determined in compile time.
In this PR, the dot prod instruction is an extended feature, and the program has to check it before.

Some pointers
 - [`HWCAP_ASIMDDP` is what you are looking for](https://www.kernel.org/doc/html/latest/arm64/elf_hwcaps.html#the-hwcaps-exposed-in-at-hwcap )
 - [What's happening under the hood](http://tomoyo.osdn.jp/cgi-bin/lxr/source/arch/arm64/kernel/cpuinfo.c#L66)

Here's the corresponding code in OpenCV (for Armv7)
https://github.com/opencv/opencv/blob/3ff1e5e6b1193421b2556683d2a6a0edda61dc31/modules/core/src/system.cpp#L560-L578
You need to populate this part and do equivalent thing in Arm v8

File: cmake/OpenCVCompilerOptimizations.cmake [hunk 6]
```diff
@@ -337,10 +341,10 @@ elseif(ARM OR AARCH64)
     endif()
     ocv_update(CPU_FP16_IMPLIES "NEON")
   else()
-    ocv_update(CPU_KNOWN_OPTIMIZATIONS "NEON;FP16")
+    ocv_update(CPU_KNOWN_OPTIMIZATIONS "NEON;FP16;DOTPROD")
```
Comment (tomoaki0705) [inline cmake/OpenCVCompilerOptimizations.cmake:344]: Same comment for here.
If you added `DOTPROD` in the list, you need to add corresponding compiler flag below.
This was tricky, since there wasn't such a "flag" in OpenCV for Armv8 till now.
You're the first one to add it.

**Review by tomoaki0705 — CHANGES_REQUESTED** [review]
Great works!
Mainly two points to fix
- [ ] Critical but not crucial, compiler error about `vtrn1q_u64`
- [ ] Critical and crucial, you've misunderstood some parts of dispatch feature. I left some pointers, and I hope it'll help you.

**Comment by Developer-Ecosystem-Engineering** [comment]
We've updated based on the feedback thanks @tomoaki0705. We will review DOTPROD in a new PR so we can disentangle it from the rest of these changes which we believe are ready at this point.

File: modules/core/include/opencv2/core/hal/intrin_neon.hpp [hunk 7]
```diff
@@ -1829,6 +2004,29 @@ inline v_int32x4 v_trunc(const v_float64x2& a)
...
 
+#if CV_NEON_AARCH64
+#define OPENCV_HAL_IMPL_NEON_TRANSPOSE4x4(_Tpvec, suffix) \
+inline void v_transpose4x4(const v_##_Tpvec& a0, const v_##_Tpvec& a1, \
+                         const v_##_Tpvec& a2, const v_##_Tpvec& a3, \
+                         v_##_Tpvec& b0, v_##_Tpvec& b1, \
+                         v_##_Tpvec& b2, v_##_Tpvec& b3) \
+{ \
+    /* -- Pass 1: 64b transpose */ \
+    _Tpvec##_t t0 = vtrn1q_##suffix##64(a0.val, a2.val); \
```
[the remark this answers is not in the document]
Comment (Developer-Ecosystem-Engineering) [reply modules/core/include/opencv2/core/hal/intrin_neon.hpp]: Fixed in the latest

**Comment by tomoaki0705** [comment]
Great work!
Just to make sure before merge, it seems that these comment lines could be removed.

https://github.com/opencv/opencv/blob/1c3f2a6ce94f6df53982f6803c462f97ec5ae89f/modules/core/test/test_intrin_utils.hpp#L585
https://github.com/opencv/opencv/blob/1c3f2a6ce94f6df53982f6803c462f97ec5ae89f/modules/core/test/test_intrin_utils.hpp#L595

**Comment by Developer-Ecosystem-Engineering** [comment]
> Great work!
> Just to make sure before merge, it seems that these comment lines could be removed.
> 
> https://github.com/opencv/opencv/blob/1c3f2a6ce94f6df53982f6803c462f97ec5ae89f/modules/core/test/test_intrin_utils.hpp#L585
> 
> 
>

Removed the commented out lines

**Comment by tomoaki0705** [comment]
Looks good to me

**Review by alalek — APPROVED** [review]
